metricflow/protocols/query_parameter.py

```diff
@@ -1,28 +1,42 @@
 from __future__ import annotations
 
-from typing import Optional, Protocol
+from typing import Optional, Protocol, Union, runtime_checkable
 
 from dbt_semantic_interfaces.type_enums import TimeGranularity
 
 from metricflow.time.date_part import DatePart
 
 
-class QueryParameterDimension(Protocol):
-    """A query parameter with a grain."""
+@runtime_checkable
+class MetricQueryParameter(Protocol):
+    """Metric requested in a query."""
 
     @property
     def name(self) -> str:
-        """The name of the item."""
+        """The name of the metric."""
         raise NotImplementedError
 
+
+@runtime_checkable
+class GroupByQueryParameter(Protocol):
+    """Generic group by parameter for queries. Might be an entity or a dimension."""
+
     @property
-    def grain(self) -> Optional[TimeGranularity]:
-        """The time granularity."""
+    def name(self) -> str:
+        """The name of the metric."""
         raise NotImplementedError
 
+
+@runtime_checkable
+class TimeDimensionQueryParameter(Protocol):  # noqa: D
```

Comment:
I don't think we need `TimeDimensionQueryParameter`. I think it would be simpler if `GroupByQueryParameter` just has the `grain` property with the `Optional[TimeGranularity]` type.

Reply:
Even though that puts the onus on all clients to use objects with invalid attributes ? I think that simplifies typing a little bit here but complicates things everywhere else. (Not sure if you read my PR description!)

Reply:
I read the PR description. I guess I'm just not understanding the burden for the client. If the object doesn't have a grain then the value would just be `None`.

Reply:
So I see the problem as this: if I'm using typed inputs as a client (like both GQL and JDBC are), then I have a `Metric` object, a `Dimension` object, an `Entity` object, and a `TimeDimension` object. If I want to pass most of those objects into an MF query, they will fail. Instead I'll have to create some new object to convert `Metric`, `Dimension`, or `Entity` objects into a new, compatible `GroupBy` object with `grain` and `date_part`, even though `grain` and `date_part` aren't relevant for those objects.

And also, on the MF side we then have to validate that a `Metric/Dimension/Entity` doesn't have a `time_granularity` or `date_part`, when the type checking should be able to do that for us.

Reply:
This is really a matter of opinion, so if you and @tlento both agree, then I'm happy to revert to that for the `group_by` param.

Reply:
> if I'm using typed inputs as a client (like both GQL and JDBC are), then I have a Metric object, a Dimension object, an Entity object, and a TimeDimension object. If I want to pass most of those objects into an MF query, they will fail. Instead I'll have to create some new object to convert Metric, Dimension, or Entity objects into a new, compatible GroupBy object with grain and date_part, even though grain and date_part aren't relevant for those objects.

This seems like a thing we should fix. How, I don't know. Maybe with this PR, maybe some other way? Or maybe it's not that big of a deal?